### ISSUE TITLE
Any downside to getting the httpClient from the HttpClientFactory in Startup.cs?

### ISSUE THREAD

**Issue by Joel-S-Black** [dotnet/runtime#68866]
Hi

My coworker brought this to my attention: [https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests) . I agree with HttpClient should be able to used against different DNS in some scenarios, but this implementation from Microsoft complicate the code. The classes have to know about the dependency of the httpclient, not just the httpclient. This disrupts the philosophy of inversion of control, since the class needs to understand how to get its dependency. 

Instead of simply registering the HttpClientFactory and passing it around, is there a negative side-effect to doing the following in Startup.cs:

```c#
services.AddHttpClient();

services.AddSingleton<HttpClient>
            (singleton => {
                var factory = singleton.GetRequiredService<IHttpClientFactory>();
                return factory.CreateClient();
            });
```

From what I can tell, this approach will eliminate refactoring for the rest of the code base, but still let the HttpClientFactory do the instance management. Also, our tests will only have to setup mocks on the HttpMessageHandler, not the HttpClientFactory itself.

**Comment by Joel-S-Black** [comment]
The github code is cleaning up my angle brackets (I hate AI), here is a screenshot of my code snippet.

![image](https://user-images.githubusercontent.com/8223950/166804490-762d2fe8-7f43-4784-bf4e-cdf24b74a6af.png)

**Comment by gfoidl** [comment]
> github code is cleaning up my angle brackets

That's how web works, no AI involved 😉 Nothing to hate here, rather have a look into the [docs about markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) how to quote your code.

> The classes have to know about the dependency of the httpclient, not just the httpclient. This disrupts the philosophy of inversion of control, since the class needs to understand how to get its dependency.

The linked article is in full conformace of IoC and DI. Only the `CatalogService` is aware of the `HttpClient`, and every consumer of that class (like `CatalogController`)  even doesn't have to know that the `CatalogService` is based on `HttpClient`. 
This is a really fine decoupling, and a huge improvement in test-ability.

> is there a negative side-effect to doing the following

Not talking about technical side-effects, rahter on architectural, then with that style you have 
* to inject HttpClient in consumers instead of typed client which hides the HttpClient from the consumer
* in tests you need to mock the HttpMessageHandler (which can be a bit cumbersome) instead of mocking the interface of the typed client which is quite easy to mock
* What if for whatever reason you don't want to use HttpClient anymore? Say you want to use gRPC solely for communication. Then you would need to adjust every consumer. With a typed client it's just this one single place to adapt, all the consumers and all the tests don't need to be touched.

So as good meant advice, re-think your approach and try to get familiar with the typed client concept. 
I'm pretty sure it will help you.

**Comment by Joel-S-Black** [comment]
Every code change needs a proveable reason where I work. 

The answer to your question on gRPC is that gRPC has not been shown to have a defined advantage over REST for my real world scenarios. I am supporting an Enterprise application where refactoring time is granted begrudgingly by those who control the budget. No services my team talks to uses http/2. The only reason I am considering making this HttpClientFactory adoption is to keep the upgrade path as short as possible for future .NET releases since my experience with Microsoft's upgrades is being as close to the book as possible. To provide color on my background and experience in this space, I survived the WebMatrix, Webforms, Silverlight, XAML, the xproj debacle, .NET Core 1.x, .NET Core 2.1's Linux issues (Workstation vs Server mode), and the NetStandard confusions, so forgive me if I am gun-shy on blindly following the next new "hip" thing Microsoft is touting. 

I am using a simplified version of the typed client, hence the need to pass the HttpClient to it. I am interested in unit-testing my typed clients so that as we make quick changes, we have unit tests to protect us in our build. I am insulted that you felt I wasn't using them. 

Your Markdown engine messed up my `` code marks. I had line breaks, since this is how I code. I'm insulted that you felt I had not read the Markdown instructions. 

So basically, you are saying as a Microsoft representative that there is no harm to my approach? Thanks.

**Comment by gfoidl** [comment]
To clarify some things:
* I mentioned gRPC only as example -- not as pro or con to REST
* the pattern used by typed client isn't a "Microsoft touting", rather a general used pattern (also in other languages / platforms)
* I read in your first comment that you are using unit tests -- I never wrote something different, just pointed out that with the typed client mocking in the tests becomes easier
* I'm no "Microsoft representative", just someone who contributes to some repositories in the dotnet-org on GH (as I'm using .NET extensively for my work, and I also "survived" the same things you mentioned)
* it's not my markdown engine, BTW for code-block you need 3 backticks ```

---

You could use the HttpClient as singleton. But there's no need to construct that single instance with the HttpClientFactory, that's pure overkill in that scenario. Just create the instance directly in the registration in DI.
But then please consider the points outlined in the article you linked on top of this issue, and the points in my first comment here.
As we have a similar long journey with .NET trust me, that the way with typed client is better in the long-term.

**Comment by Joel-S-Black** [comment]
After reading your last statement @gfoidl , I apologize for thinking you were a Microsoft person answering my question. I also want to apologize for omitting the following link: [https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-6.0](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-6.0) . I'd also like to apologize for thinking it was your Markdown engine (since I thought you worked for Microsoft, who owns Github). 

In this example, the HttpClientFactory, not the HttpClient is injected, which is why I was motivated to post this issue. Another  downside to how they are creating their typed client is it is Transient; in the past, getting the access token was expensive for the other service, and we had to re-use them as Scoped. Hence, my approach to create a new Client while letting the factory handle the lifetime. 

Will a Microsoft coder who has been dealing with the fallout the iterations of HttpClient (usings, then new, then singleton, now factory) weigh in on my question?

**Comment by CarnaViire** [comment]
Hi @Joel-S-Black,

First of all, please be respectful to others, regardless of whether the person works in Microsoft or not. Also, it would be most helpful if you could focus on the topic related to .NET Core/.NET 5+ in this repo.

Returning to the question in hand, short answer is:
Yes you will have problems with that approach. The resulting singleton `HttpClient` will have all the same problems as a plain static `HttpClient` -- you may notice the loss of DNS changes.

`HttpClientFactory` is no magic wand that will give you a "fixed" `HttpClient`. The factory solves socket exhaustion problem by pooling, reusing and recreating `HttpClientHandler`s. `HttpClient`s created by the factory are expected to be short-lived, and it is OK because handlers are reused. After `HttpClient` is created, `HttpClientFactory` has no control over it anymore, so it plainly cannot handle it's lifetime.

I understand that the documentation is in a bad shape right now, so it must be very frustrating for you. (We do have plans for improving the docs, e.g. https://github.com/dotnet/dotnet-api-docs/issues/7715)

If you use .NET Core/.NET 5+, we recommend using a static/singleton `HttpClient` *with PooledConnectionLifetime set* to a desired interval depending on expected DNS changes (e.g. you can use 2 mins as `HttpClientFactory` does by default):

```c#
new HttpClient(new SocketsHttpHandler
{
    PooledConnectionLifetime = TimeSpan.FromMinutes(2) //recreate every 2 minutes
});
```

If you wish to register your handler separately to be able to mock it, it could be
```c#
services.AddSingleton<HttpClientHandler>(sp => new SocketsHttpHandler { PooledConnectionLifetime = TimeSpan.FromMinutes(2) });
services.AddSingleton<HttpClient>(sp => new HttpClient(sp.GetRequiredService<HttpClientHandler>()));
```

This would solve both socket exhaustion and DNS changes problems without adding an overhead of `HttpClientFactory`.

If you require the full configurability `HttpClientFactory` offers, though, like having multiple differently configured `HttpClient`s for different use cases, you are welcome to use Typed client approach, though please be aware, that Typed clients are also intended to be short-lived, and you may shoot yourself in a foot if you inject it into a sigleton. See e.g. https://github.com/dotnet/runtime/issues/64034

Please let me know if this helps.

cc @karelz

**Comment by Joel-S-Black** [comment]
@CarnaViire,

> Also, it would be most helpful if you could focus on the topic related to .NET Core/.NET 5+ in this repo.

Will you clarify this statement? I am referring to articles published in the past month by the company that hosts this repository since another issue was raised about this subject in the same repo; what prompted this request for information is a pending upgrade to .NET 6; currently, we are on .NET Core 3.1 and are looking into best practices. 

> If you use .NET Core/.NET 5+, we recommend using a static/singleton HttpClient with PooledConnectionLifetime set to a desired interval depending on expected DNS changes (e.g. you can use 2 mins as HttpClientFactory does by default).

Why is this guidance different from the most recent guidance (April 2022) published by Microsoft on the subject? Has HttpClient changed in .NET 5 from .NET 3.1?

**Comment by wfurt** [comment]
Sorry to chime in late but I feel lot of the confusion and issues comes from the fact that users do not understand how it actually works and we do not have great documentation. Partially for historical reasons and lack of time and partially because it is internal implementation detail. Also some of it depends on specific use pattern so there not simple single solution that fits everything. 

in .NET Framework the connection pool was linked to ServicePointManager and HttpClient was implemented on top of HttpWebRequest. So disposing HttpClient did not impact connection management.

In .NET Core, the connection pool is linked to HttpClient's underlying HttpHandler. So when HttpClient is disposed it will dispose all previously used connections. That forces new connection next time you need to talk to the same server and there is performance penalty as well it needs new TCP port. If the rate of request is high or if there are any firewall limitations that can exhaust available range because of default TCP cleanup timers.   

So tho work efficiently as possible it is good to reuse HttpClient. Either as singleton or via Factory as @CarnaViire suggested.  

The DNS issue was always somewhat mysterious to me. If you for example move server to new IP and update DNS it will work as the existing connections would see error and there is new DNS lookup every time HttpClient needs connection (e.g. there is no magic DNS caching behind the scenes). The only once case that is not handled well is case when existing server decides to stop handling request for given site silently e.g. without closing connections first. This is where  the `PooledConnectionLifetime` comes in to force recreation every so often (mostly without real need)

So depending on your use single static HttpClient  _may_ work for you just fine @Joel-S-Black.

**Comment by CarnaViire** [comment]
Closing as answered
